org.eclipse.jdt.core/codeassist/org/eclipse/jdt/internal/codeassist/CompletionEngine.java

```diff
@@ -12332,32 +12352,52 @@ private void findTypesFromExpectedTypes(char[] token, Scope scope, ObjectVector
 						}
 
 						if (proposeType &&
-								(!this.assistNodeIsConstructor || dimensions > 0 ||
+								(!this.assistNodeIsConstructor || isArrayCompletion ||
 										!allowingLongComputationProposals ||
 										hasStaticMemberTypes(refBinding, scope.enclosingSourceType() ,this.unitScope)) ||
 										hasArrayTypeAsExpectedSuperTypes()) {
 							this.noProposal = false;
 							if(!this.requestor.isIgnored(CompletionProposal.TYPE_REF)) {
-								InternalCompletionProposal proposal =  createProposal(CompletionProposal.TYPE_REF, this.actualCompletionPosition);
-								proposal.setDeclarationSignature(packageName);
-								proposal.setSignature(getSignature(refBinding));
-								proposal.setPackageName(packageName);
-								proposal.setTypeName(typeName);
-								proposal.setCompletion(completionName);
-								proposal.setFlags(refBinding.modifiers);
-								proposal.setReplaceRange(this.startPosition - this.offset, this.endPosition - this.offset);
-								proposal.setTokenRange(this.tokenStart - this.offset, this.tokenEnd - this.offset);
-								proposal.setRelevance(relevance);
-								proposal.setAccessibility(accessibility);
-								proposal.setArrayDimensions(dimensions);
-								this.requestor.accept(proposal);
-								if(DEBUG) {
-									this.printDebug(proposal);
+								char[] signature = getSignature(refBinding);
+								boolean found = false;
+								// check if we already found this completion before. if we have found then change it's
+								// dimension and relevance to match against expected type.
+								for (int j = 0; j < proposalsFound.size(); j++) {
```

Comment:
Any idea if this can cause trouble due to quadratic complexity?

Other than this, the change looks fine now.

Reply:
Agree, did some improvements by eliminating the need for altering the already found completions, while also eliminating the duplicates. Lets see how the tests react.